mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -202,25 +242,33 @@ def _azure_upload_file(self, credentials, local_file, artifact_path):
         try:
             headers = self._extract_headers_from_credentials(credentials.headers)
             uploading_block_list = []
-            for chunk in yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE):
-                # Base64-encode a UUID, producing a UTF8-encoded bytestring. Then, decode
-                # the bytestring for compliance with Azure Blob Storage API requests
-                block_id = base64.b64encode(uuid.uuid4().hex.encode()).decode("utf-8")
-                try:
-                    put_block(credentials.signed_uri, block_id, chunk, headers=headers)
-                except requests.HTTPError as e:
-                    if e.response.status_code in [401, 403]:
-                        _logger.info(
-                            "Failed to authorize request, possibly due to credential expiration."
-                            " Refreshing credentials and trying again..."
+            futures = {}
+            with self.thread_pool as executor:
+                for chunk_index, chunk in enumerate(
+                    yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE)
+                ):
+                    upload_future = executor.submit(
+                        self._azure_upload_chunk,
+                        credentials=credentials,
+                        chunk=chunk,
+                        local_file=local_file,
+                        artifact_path=artifact_path,
+                    )
+                    futures[upload_future] = chunk_index
+
+                    # Avoid loading too many chunks into memory
+                    if len(futures) < 5:
+                        continue
+
+                    results, errors = self._complete_futures(futures)
```

Comment:
With this approach, we're blocked on the slowest future before continuing the upload. Is this how libraries like boto3 do it? If not, can we copy the boto3 approach?

Reply:
I'm reading https://github.com/boto/s3transfer/blob/f225103e69f2e2b61e8933ad816519229afaf088/s3transfer/upload.py#L651 to see what boto3 does.

Reply:
Doing some print injection in the s3transfer source code to understand what boto3 does.

Reply:
In boto3, each future has a `ReadFileChunk` object:

https://github.com/boto/s3transfer/blob/93fb7d1c758c087bc7c87b9e45166b2b741aa408/s3transfer/utils.py#L398

As the name indicates, this object can chunk-read bytes from the file to upload.